File: helloworld.go
//-----------------------------------------------------------------------------------------
// EXAMPLE
//
//	find ~/trash/ 	| go run ~/github/templates.git/helloworld.go
//	cat ~/.zshrc 	|  go run ~/github/templates.git/helloworld.go
//
// COMPILE TO NATIVE
//
//	# this will embed everything, as of 2020
//	go build helloworld.go
//
// DATE
//
//	2022 Jan 20
//

//-----------------------------------------------------------------------------------------

package main

import (
	"bufio"
	"fmt"
	// I forgot - why is the builtin "flag" package not good enougH?
	"github.com/pborman/getopt"
	"io"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

var counts = make(map[string]int)

// golang MUST have a main function (unlike python)
func main() {

	///
	/// 5) CLI options
	///
	optName := getopt.StringLong("name", 'n', "", "Your name")
	optHelp := getopt.BoolLong("help", 0, "Help")
	getopt.Parse()

	if *optHelp {
		getopt.Usage()
		os.Exit(0)
	}
	fmt.Println("Hello " + *optName)

	///
	/// 1) Loop over stdin
	///
	in := bufio.NewReader(os.Stdin)
	for {
		s, err := in.ReadString('\n')
		if err != nil {
			// io.EOF is expected, anything else
			// should be handled/reported
			if err != io.EOF {
				log.Fatal(err)
			}
			break
		}
		// Do something with the line of text
		// in string variable s.
		_ = s
		p := strings.TrimSpace(s)

		//
		// 3) Parse file path
		//

		if file, err := os.Stat(p); !os.IsNotExist(err) {

			// check if it's a file or a dir
			switch i, err := os.Stat(p); {
			case err != nil:
				fmt.Println(err)
			case i.IsDir():
				fmt.Println("	", p, "is a directory")
			default:
				fmt.Println("	", p, "is a file")
			}
			// get absolute path
			abs, _ := filepath.Abs(p)
			fmt.Println("	absolute: " + abs)

			///
			/// 6) Call a shell program instead
			///
			cmd := exec.Command("dirname", p)
			out, err := cmd.CombinedOutput()
			if err != nil {
				log.Fatal(err, out)
			}

			cmd2 := exec.Command("basename", p)
			out2, err2 := cmd2.CombinedOutput()
			if err2 != nil {
				log.Fatal(err2)
			}
			dirname := strings.Trim(string(out2), "\n")
			//			fmt.Println("	DIRNAME: ", dirname)

			///
			/// 5) dictionary
			///
			_, exist := counts[dirname]
			if !exist {
				counts[dirname] = 1
			} else {
				counts[dirname]++
			}

			// We don't need to check if it's a folder to recurse into. Stdin will ensure that
			// we recurse.
			if counts[dirname] <= 5 {
				fmt.Println("", p, "\n")
			}

		} else {
			///
			/// 1) Print to stdout
			///
			if file == nil {
				fmt.Print("from go: ", p, "\n")
			}

			///
			/// 2) Regex capture groups extracted and read separately (not used directly
			/// in a substitution) will cover more scenarios.
			///
			regex := "^\\s*([0-9]+)*\\s*DOCUMENT_FREQUENCY_TOTAL..(.*)\n"
			r := regexp.MustCompile(regex)
			line := s
			elem := r.FindStringSubmatch(line)

			if len(elem) == 0 {
				// no match
				continue
			}

			// elem[0] is the entire line
			for i := 1; i < len(elem); i++ {
				fmt.Print(elem[i])
				fmt.Println()
			}
		}
	}
}

File: helloworld.js
//-----------------------------------------------------------------------------------------
// EXAMPLE
//
//	find ~/trash/	| node helloworld.js -n 4
//
// 	cat ~/.zshrc	| node helloworld.js
//
//
//	npm install command-line-args --save

// DATE
//
//	2022 Jan 23
//
// AUTHOR
//
//	ss401533@gmail.com
//
//-----------------------------------------------------------------------------------------


const readline		= require('readline');


	///
	/// 5) CLI options
	///

var stdio = require('stdio');
var getopt = stdio.getopt;


const options = getopt({
	num_max: {key: 'n', default : 5, description: 'If the project is finished'}
});


const optionDefinitions = [
  {
    name: 'help',
    alias: 'h',
    type: Boolean,
    description: 'Display this usage guide.'
  },
  { name: 'verbose', alias: 'v', type: Boolean },
  { name: 'src', type: String, multiple: true, defaultOption: true },
  { name: 'num-max', alias: 'n', type: Number, defaultValue : 5}
]
//const options2 = commandLineArgs(optionDefinitions)



	///
	/// 1) Loop over stdin
	///
var rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
	terminal: false
});



rl.on('line', function(line){
//	console.log(options2.n);
//	console.log(ops.kaka);
	console.log(options);
	console.log('added: ' + line);
//	if (ops.kaka && ops.check) {
//	    console.log(ops.kaka + ops.check[0]);
//	}
})
